### ISSUE TITLE
Naming of type aliases in Prelude `*/fold`

### ISSUE THREAD

**Issue by Profpatsch** [dhall-lang/dhall-lang#731]
The `*/fold` have a peculiar choice of naming of the result type:

```
Natural/fold : Natural → ∀(natural : Type) → ∀(succ : natural → natural) → ∀(zero : natural) → natural
List/fold : ∀(a : Type) → List a → ∀(list : Type) → ∀(cons : a → list → list) → ∀(nil : list) → list
```

Considering that it’s not required to make the return type a `List`/`Optional`, it’s kind of confusing to name the type variables `list`/`optional`.

Why don’t we change it to

```
Natural/fold : Natural → ∀(result : Type) → ∀(succ : result → result) → ∀(zero : result) → result
List/fold : ∀(a : Type) → List a → ∀(result : Type) → ∀(cons : a → result → result) → ∀(nil : result) → result
```

**Comment by sjakobi** [comment]
I guess these type variables were chosen to point out the correspondence between the `*/build` and the `*/fold` functions. E.g.

```
List/build : ∀(a : Type) → (∀(list : Type) → ∀(cons : a → list → list) → ∀(nil : list) → list) → List a
List/fold : ∀(a : Type) → List a → ∀(list : Type) → ∀(cons : a → list → list) → ∀(nil : list) → l
```

I agree that the types of the `*/fold` functions by themselves are confusing – renaming the variables to `result` would help!

**Comment by Gabriel439** [comment]
Yeah, the function arguments are named after what they replace.  The first argument replaces all occurrences of the list's type.  The second argument replaces all occurrences of the "cons" constructor (if the list were a linked list like in Haskell), and the third argument replaces all occurrences of the "nil" constructor.

It also has the nice property that it is symmetric with the `List/build` function, like @sjakobi mentioned.

I think that if we change the names then we should change all three for consistency, because it doesn't make sense to me to name the two other arguments `cons` and `nil` without naming the first argument `list`.  For example, if we replace `list` with `result` it won't be clear why the `cons` argument has type `a → result → result` (or even why it is named `cons`).

However, I still personally prefer keeping the names the way they are.
